### PR TITLE
[#IOCIT-344] Add revokePubKeys Queue handler

### PR DESCRIPTION
**Pull request by AleDore**
#### List of Changes

- Add `RevokePubKey` handler
- Add `lollipopKeys` utility
- Add unit tests
- Fix integration tests
#### Motivation and Context

This is required to provide an async mechanism that could accomplish pubKeys revocation.
#### How Has This Been Tested?

Unit tests
#### Types of changes

- [ ] Chore (nothing changes by a user perspective)
- [ ] Bug fix (non-breaking change which fixes an issue)
- [ ] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to not work as expected)

#### Checklist:

- [ ] My change requires a change to the documentation.
- [ ] I have updated the documentation accordingly.

### CODE REVIEW

File: HandlePubKeyRevoke/handler.ts [hunk 1]
```diff
@@ -0,0 +1,164 @@
...
+                  RA.map(lollipopKey =>
+                    lollipopKeysModel.upsert({
+                      ...lollipopKey,
+                      status: PubKeyStatusEnum.REVOKED
+                    })
+                  ),
+                  RA.sequence(TE.ApplicativeSeq),
+                  TE.mapLeft(_ =>
+                    toTransientFailure(
+                      Error("Cannot perform upsert CosmosDB")
```
Comment (fabriziopapi) [inline HandlePubKeyRevoke/handler.ts:125]:
```suggestion
                  TE.mapLeft(err =>
                    toTransientFailure(
                      Error(
                        `Cannot perform upsert CosmosDB: ${JSON.stringify(err)}`
                      )
```

File: HandlePubKeyRevoke/handler.ts [hunk 2]
```diff
@@ -0,0 +1,164 @@
...
+    RevokeAssertionRefInfo.decode,
+    TE.fromEither,
+    TE.mapLeft(flow(errorsToError, e => toPermanentFailure(e)())),
+    TE.chain(revokeAssertionRefInfo =>
+      pipe(
+        lollipopKeysModel.findLastVersionByModelId([
+          revokeAssertionRefInfo.assertion_ref
+        ]),
+        TE.mapLeft(_ =>
+          toTransientFailure(Error("Cannot perform find on CosmosDB"))()
```
Comment (fabriziopapi) [inline HandlePubKeyRevoke/handler.ts:106]:
```suggestion
        TE.mapLeft(err =>
          toTransientFailure(
            Error(`Cannot perform find on CosmosDB: ${JSON.stringify(err)}`)
          )()
```

File: HandlePubKeyRevoke/index.ts [hunk 3]
```diff
@@ -0,0 +1,34 @@
...
+import { initTelemetryClient } from "../utils/appinsights";
+import { getConfigOrThrow } from "../utils/config";
+import { cosmosdbInstance } from "../utils/cosmosdb";
+import { Failure } from "../utils/errors";
+import { handleRevoke } from "./handler";
+
+const config = getConfigOrThrow();
+
+const lollipopKeysModel = new LolliPOPKeysModel(
+  cosmosdbInstance.container("PIPPO")
```
Comment (gquadrati) [inline HandlePubKeyRevoke/index.ts]: Pippo? :D

File: HandlePubKeyRevoke/handler.ts [hunk 4]
```diff
@@ -0,0 +1,168 @@
...
+  NotPendingLolliPopPubKeys
+} from "../model/lollipop_keys";
+import { PubKeyStatusEnum } from "../generated/definitions/internal/PubKeyStatus";
+import { JwkPubKeyHashAlgorithm } from "../generated/definitions/internal/JwkPubKeyHashAlgorithm";
+import { getAllAssertionsRef } from "../utils/lollipopKeys";
+
+/**
+ * Based on a previous retrieved LollipopPubKey that match with assertionRef retrieved on queue
+ * this function extracts all lollipopPubKeys to be revoked including master key
+ *
```
Comment (gquadrati) [inline HandlePubKeyRevoke/handler.ts:30]: Explain array order

File: HandlePubKeyRevoke/handler.ts [hunk 5]
```diff
@@ -0,0 +1,168 @@
...
+                  )()
+                )
+              ),
+              TE.chainEitherK(
+                flow(
+                  NotPendingLolliPopPubKeys.decode,
+                  E.mapLeft(() =>
+                    toPermanentFailure(
+                      Error("Cannot decode a VALID master lollipopPubKey")
+                    )()
```
Comment (gquadrati) [inline HandlePubKeyRevoke/handler.ts:74]:
```suggestion
                    toTransientFailure(
                      Error("Cannot decode a VALID master lollipopPubKey")
                    )()
```

File: HandlePubKeyRevoke/handler.ts [hunk 6]
```diff
@@ -0,0 +1,168 @@
...
+                  )
+                )
+              ),
+              TE.map(validMasterLollipopPubKeys => [
+                validMasterLollipopPubKeys,
+                notPendingLollipopPubKeys
+              ])
```
Comment (gquadrati) [inline HandlePubKeyRevoke/handler.ts:81]: Switch to struct

File: HandlePubKeyRevoke/handler.ts [hunk 7]
```diff
@@ -0,0 +1,168 @@
...
+              extractPubKeysToRevoke(lollipopKeysModel, masterAlgo),
+              TE.chainW(
+                flow(
+                  RA.map(lollipopKey =>
+                    lollipopKeysModel.upsert({
+                      ...lollipopKey,
+                      status: PubKeyStatusEnum.REVOKED
+                    })
+                  ),
+                  RA.sequence(TE.ApplicativeSeq),
```
Comment (gquadrati) [inline HandlePubKeyRevoke/handler.ts:124]:
```suggestion
                  RA.sequence(TE.ApplicativePar),
```

File: HandlePubKeyRevoke/handler.ts [hunk 8]
```diff
@@ -0,0 +1,168 @@
...
+        tagOverrides: { samplingEnabled: String(isTransient) }
+      });
+      context.log.error(error);
+      if (isTransient) {
+        // Trigger a retry in case of temporary failures
+        throw new Error(error);
+      }
+      return err;
+    }),
+    TE.map(constVoid),
```
Comment (gquadrati) [inline HandlePubKeyRevoke/handler.ts:165]:
```suggestion
```

File: HandlePubKeyRevoke/__tests__/handler.test.ts [hunk 9]
```diff
@@ -0,0 +1,465 @@
...
+      expect.objectContaining({
+        kind: "PERMANENT",
+        reason: expect.stringContaining(
+          "Cannot decode a VALID master lollipopPubKey"
+        )
+      })
+    );
+  });
+
+  it("GIVEN a valid revoke message that does not match master key WHEN master lollipopPubKey is retieved THEN it should upsert to REVOKED all pubKeys", async () => {
```
Comment (gquadrati) [inline HandlePubKeyRevoke/__tests__/handler.test.ts:421]:
```suggestion
  it("GIVEN a valid revoke message that does not match master key WHEN master lollipopPubKey is retrieved THEN it should upsert to REVOKED all pubKeys", async () => {
```

**Review by gquadrati — CHANGES_REQUESTED** [review]

File: utils/lollipopKeys.ts [hunk 10]
```diff
@@ -0,0 +1,90 @@
...
+  JwkPubKeyHashAlgorithm,
+  JwkPubKeyHashAlgorithmEnum
+} from "../generated/definitions/internal/JwkPubKeyHashAlgorithm";
+import { AssertionRefSha512 } from "../generated/definitions/internal/AssertionRefSha512";
+import { AssertionRef } from "../generated/definitions/internal/AssertionRef";
+import { AssertionRefSha384 } from "../generated/definitions/internal/AssertionRefSha384";
+import { AssertionRefSha256 } from "../generated/definitions/internal/AssertionRefSha256";
+import { assertNever } from "./errors";
+
+export const AssertionsRef = t.intersection([
```
Comment (gquadrati) [inline utils/lollipopKeys.ts:19]: rename to `AssertionRefByType`

File: utils/errors.ts [hunk 11]
```diff
@@ -0,0 +1,83 @@
...
+  const error = TransientFailure.is(err)
+    ? `${logPrefix}|TRANSIENT_ERROR=${err.reason}`
+    : `${logPrefix}|FATAL|PERMANENT_ERROR=${err.reason}`;
+  trackException(telemetryClient, {
+    exception: new Error(error),
+    properties: {
+      detail: err.kind,
+      fatal: PermanentFailure.is(err).toString(),
+      isSuccess: "false",
+      name: `cgn.exception.${logPrefix}.failure`
```
Comment (gquadrati) [inline utils/errors.ts:75]: cgn?

File: HandlePubKeyRevoke/handler.ts (hunk 4 above)
Comment (AleDore) [reply HandlePubKeyRevoke/handler.ts:36]: Modified

**Pull request merged**